tensorpipe/test/channel/channel_test.cc

```diff
@@ -15,513 +15,568 @@
-
-        // Allocate a non-empty vector so that its .data() pointer is non-null.
-        std::vector<uint8_t> data(1);
-
-        // Perform recv and wait for completion.
-        auto descriptor = peers_->recv(PeerGroup::kClient);
-        std::future<Error> recvFuture =
-            recvWithFuture(channel, descriptor, CpuTensor{data.data(), 0});
-        Error recvError = recvFuture.get();
-        EXPECT_FALSE(recvError) << recvError.what();
-
-        peers_->done(PeerGroup::kClient);
-        peers_->join(PeerGroup::kClient);
-
-        ctx->join();
-      });
-}
-
-TEST_P(ChannelTest, contextIsNotJoined) {
+  void server(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<Context<TTensor>> ctx =
+        this->helper_->makeContext("server");
+    auto channel = ctx->createChannel(std::move(conn), Endpoint::kListen);
+
+    // Perform send and wait for completion.
+    std::future<std::tuple<Error, TDescriptor>> descriptorFuture;
+    std::future<Error> sendFuture;
+    std::tie(descriptorFuture, sendFuture) = sendWithFuture(
+        channel,
+        TTensor{
+            .ptr = nullptr,
+            .length = 0,
+        });
+    Error descriptorError;
+    TDescriptor descriptor;
+    std::tie(descriptorError, descriptor) = descriptorFuture.get();
+    EXPECT_FALSE(descriptorError) << descriptorError.what();
+    this->peers_->send(PeerGroup::kClient, descriptor);
+    Error sendError = sendFuture.get();
+    EXPECT_FALSE(sendError) << sendError.what();
+
+    this->peers_->done(PeerGroup::kServer);
+    this->peers_->join(PeerGroup::kServer);
+
+    ctx->join();
+  }
+
+  void client(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<Context<TTensor>> ctx =
+        this->helper_->makeContext("client");
+    auto channel = ctx->createChannel(std::move(conn), Endpoint::kConnect);
+
+    // Perform recv and wait for completion.
+    auto descriptor = this->peers_->recv(PeerGroup::kClient);
+    std::future<Error> recvFuture = recvWithFuture(
+        channel,
+        descriptor,
+        TTensor{
+            .ptr = nullptr,
+            .length = 0,
+        });
+    Error recvError = recvFuture.get();
+    EXPECT_FALSE(recvError) << recvError.what();
+
+    this->peers_->done(PeerGroup::kClient);
+    this->peers_->join(PeerGroup::kClient);
+
+    ctx->join();
+  }
+};
+
+// Call send and recv with a length of 0 but a non-null pointer.
+template <typename TTensor>
+class EmptyTensorTest : public ChannelTest<TTensor> {
+  void server(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<Context<TTensor>> ctx =
+        this->helper_->makeContext("server");
+    auto channel = ctx->createChannel(std::move(conn), Endpoint::kListen);
+
+    // Allocate a non-empty vector so that its .data() pointer is non-null.
+    std::vector<uint8_t> data(1);
+    DataWrapper<TTensor> wrappedData(data);
+    TTensor tensor = wrappedData.tensor();
+    tensor.length = 0;
```

Comment:
Similarly non-generic.

Reply:
It looks like one comment I though I had made before has disappeared, I'm rewriting it now.